pkg/controllers/management/clusterupstreamrefresher/eks_upstream_spec.go

```diff
@@ -0,0 +1,136 @@
+	sess, eksService, err := ekscontroller.StartAWSSessions(secretsCache, *cluster.Spec.EKSConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	clusterState, err := eksService.DescribeCluster(
+		&eks.DescribeClusterInput{
+			Name: aws.String(cluster.Spec.EKSConfig.DisplayName),
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	ngs, err := eksService.ListNodegroups(
+		&eks.ListNodegroupsInput{
+			ClusterName: aws.String(cluster.Spec.EKSConfig.DisplayName),
+		})
+
+	// gather upstream node groups states
+	var nodeGroupStates []*eks.DescribeNodegroupOutput
+	var errs []string
+	for _, ngName := range ngs.Nodegroups {
+		ng, err := eksService.DescribeNodegroup(
+			&eks.DescribeNodegroupInput{
+				ClusterName:   aws.String(cluster.Spec.EKSConfig.DisplayName),
+				NodegroupName: ngName,
+			})
+		if err != nil {
+			return nil, err
+		}
+
+		nodeGroupStates = append(nodeGroupStates, ng)
+		var nodeGroupMustBeDeleted string
+		if len(ng.Nodegroup.Health.Issues) != 0 {
+			var issueMessages []string
+			for _, issue := range ng.Nodegroup.Health.Issues {
+				issueMessages = append(issueMessages, aws.StringValue(issue.Message))
+				if !ekscontroller.NodeGroupIssueIsUpdatable(aws.StringValue(issue.Code)) {
+					nodeGroupMustBeDeleted = ": node group cannot be updated, must be deleted and recreated"
+				}
+			}
+			errs = append(errs, fmt.Sprintf("health error for node group [%s] in cluster [%s]: %s%s",
+				aws.StringValue(ng.Nodegroup.NodegroupName),
+				cluster.Name,
+				strings.Join(issueMessages, "; "),
+				nodeGroupMustBeDeleted,
+			))
+		}
+	}
+
+	upstreamSpec, _, err := ekscontroller.BuildUpstreamClusterState(cluster.Spec.DisplayName, cluster.Status.EKSStatus.ManagedLaunchTemplateID, clusterState, nodeGroupStates, ec2.New(sess), false)
+	if err != nil {
+		// If we get an error here, then syncing is broken
+		return nil, err
+	}
+
+	upstreamSpec.DisplayName = cluster.Spec.EKSConfig.DisplayName
+	upstreamSpec.Region = cluster.Spec.EKSConfig.Region
+	upstreamSpec.AmazonCredentialSecret = cluster.Spec.EKSConfig.AmazonCredentialSecret
+	upstreamSpec.Imported = cluster.Spec.EKSConfig.Imported
+	upstreamSpec.Subnets = cluster.Spec.EKSConfig.Subnets
+	upstreamSpec.SecurityGroups = cluster.Spec.EKSConfig.SecurityGroups
+	upstreamSpec.ServiceRole = cluster.Spec.EKSConfig.ServiceRole
+
+	if len(errs) != 0 {
+		// If there are errors here, we can still sync, but there are problems with the nodegroups that should be reported
+		err = fmt.Errorf("error for cluster [%s]: %s",
+			cluster.Name,
+			strings.Join(errs, "\n"))
+	}
+
+	return upstreamSpec, err
+}
+
+// MigrateEksRefreshCronSetting migrates the deprecated eks-refresh-cron setting to new
+// setting only if default setting was changed
+// This function will be run only once during startup by pkg/multiclustermanager/app.go
+func MigrateEksRefreshCronSetting(wContext *wrangler.Context) {
```

Comment:
The migration should only happen once unless there is an error. To accomplish this the deprecated setting needs to have an annotation added showing the migration was completed. This will serve a 2 purposes: 1. It gives a fast exit - if the old setting has the anno no more processing is required. 2. It shows that rancher did do the migration which is important for testing and from a users/support perspective showing rancher did its job of attempting a 1 time migration.

Reply:
Annotation was done

Reply:
Oh I thought we were just going to check values but this is good too.